### PR TITLE
Trimming

### PR DESCRIPTION
**Pull request by mbannick**
- Builds on the #5 gradient PR to implement trimming. I suppose we don't need the analytical gradient for this, actually.

### CODE REVIEW

File: notebooks/helpers.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+        if ineff_dist == 'half-normal':
+            self.rvs = halfnorm.rvs
+        elif ineff_dist == 'exponential':
+            self.rvs = expon.rvs
+        else:
+            raise RuntimeError("Inefficiency distribution must be half-normal or exponential")
+
+    def simulate(self, n: int = 1, **kwargs):
+        np.random.seed(365)
```
Comment (zhengp0) [inline notebooks/helpers.py]: will be nice if we could pass `seed` as a key word argument

File: src/sfma/api.py [hunk 2]
```diff
@@ -44,6 +49,11 @@ def __init__(self,
         self.l_linear = l_linear
         self.include_intercept = include_intercept
         self.include_gamma = include_gamma
+        self.pct_trimming = pct_trimming
+        self.constr_grid_num = constr_grid_num
+
+        if pct_trimming > 1.0:
```
Comment (zhengp0) [inline src/sfma/api.py]: Also good to check the case less than 0. `if pct_trimming < 0 or pct_trimming > 1`

File: src/sfma/api.py [hunk 3]
```diff
@@ -134,8 +145,56 @@ def __init__(self,
         self.inefficiencies = np.zeros(self.data.num_obs)
 
     def fit(self, **kwargs):
-        self.solver.fit(x_init=self.x_init, data=self.data, **kwargs)
+        options = kwargs.get('options', {'solver_options': {}})
+        self.trimming.fit(x_init=self.x_init, data=self.data, n=len(self.data.obs),
```
Comment (zhengp0) [inline src/sfma/api.py]: Will be nice if we could overwrite `self.x_init`, really like what you did for `options` we could follow that do `x_init = kwargs.get("x_init", self.x_init)`

File: src/sfma/models/marginal.py [hunk 4]
```diff
@@ -70,15 +72,23 @@ def remat(self) -> ndarray:
...
         gamma = np.sqrt(variables[1]**2)
         eta = np.sqrt(variables[2]**2)
         return beta, gamma, eta
 
+    @property
+    def w(self):
+        return self._w
+
+    @w.setter
+    def w(self, weights):
```
Comment (zhengp0) [inline src/sfma/models/marginal.py]: To take advantage of `setter` we could add checks of the weights, to enforce it between 0 and 1.
```
if any(map(lambda x: x < 0 or x > 1), weights):
    raise ValueError
```

File: src/sfma/models/marginal.py [hunk 5]
```diff
@@ -90,20 +100,18 @@ def objective(self, x: ndarray, data: Data) -> float:
...
-        return np.mean(0.5*r**2/v + 0.5*np.log(v) - log_erfc(z))
+        return 0.5 * r ** 2 / v + 0.5 * np.log(v) - log_erfc(z)
 
-    def gradient(self, x: ndarray, data: Data) -> ndarray:
-        """
-        Computes the gradient.
+    def objective(self, x: ndarray, data: Data) -> float:
+        obj = self._objective(x=x, data=data)
+        if self.w is not None:
+            obj = self.w.dot(obj)
```
Comment (zhengp0) [inline src/sfma/models/marginal.py:127]: !Inconsistency with the gradient below, here we didn't take the mean. When `self.w is not None`, `self.w.dot(obj)` will be a scalar, the mean below won't do anything. Imagine the case when `self.w` are a vector with all ones.

File: src/sfma/models/marginal.py [hunk 6]
```diff
@@ -117,13 +125,27 @@ def gradient(self, x: ndarray, data: Data) -> ndarray:
...
+        :param w: optional weights
+        :return:
+        """
+        grad = self._gradient(x=x, data=data)
+        if self.w is not None:
+            grad = self.w * grad
+        grad = np.sum(grad, axis=1)
 
         # Take the average because the objective
         # is the mean rather than the sum
```
Comment (zhengp0) [inline src/sfma/models/marginal.py:170]: !Inconsistency with objective, here gradient is the mean, but the objective function is the sum. Might be the reason why the algorithm have a hard time to converge :)

**Review by zhengp0 — COMMENTED** [review]
Thanks @mbannick for working on this! Left some minor comments. One issue is that the objective function and gradient don't match. It will be best if we use a gradient test to check this. Let me know if there is anything confusing or anything I can help with!

File: notebooks/helpers.py (hunk 1 above)
Comment (mbannick) [reply notebooks/helpers.py]: 87a0ac2

File: src/sfma/api.py (hunk 2 above)
Comment (mbannick) [reply src/sfma/api.py]: 77efaf0

File: src/sfma/api.py (hunk 3 above)
Comment (mbannick) [reply src/sfma/api.py]: 1b6d642 -- switched to `kwargs.pop` :)

File: src/sfma/models/marginal.py (hunk 4 above)
Comment (mbannick) [reply src/sfma/models/marginal.py]: I like that! 5a28339

**Pull request merged**